File: Season-1/Level-1/code.py
'''
Welcome to Secure Code Game Season-1/Level-1!

Follow the instructions below to get started:

1. tests.py is passing but code.py is vulnerable
2. Review the code. Can you spot the bug?
3. Fix the code but ensure that tests.py passes
4. Run hack.py and if passing then CONGRATS!
5. If stuck then read the hint
6. Compare your solution with solution.py
'''

from collections import namedtuple
from decimal import Decimal

Order = namedtuple('Order', 'id, items')
Item = namedtuple('Item', 'type, description, amount, quantity')

def my_validorder(order: Order):
    """
    hack.py : 2/3
    tests.py: oooxo
    """
    net = 0

    for item in order.items:
        if item.type == 'payment':
            # 1. hack.pyからpaymentにマイナス値があることを想定する必要
            #    => オーバーフローチェック時にはabs()を使う
            net += Decimal(item.amount)
        elif item.type == 'product':
            # 2. item.quantityがint型か確認する必要 (Decimal * floatが不可)
            # 3. 購入費の上限を設定して、適切にquantityとamountの上限を検査する必要 (Overflow対策)
            net -= Decimal(item.amount) * item.quantity
        else:
            return "Invalid item type: %s" % item.type

    # round()だと間違い。
    # netの値を最初からDecimal(str())の型でデータ保持・演算を行う
    # なお、この方法だと`1e+19`のような表現をされたときにunderflowの懸念
    # が残るので、item.amountの上限を検査する等の他の対策も必要
    if round(net) != 0:
        return "Order ID: %s - Payment imbalance: $%0.2f" % (order.id, net)
    else:
        return "Order ID: %s - Full payment received!" % order.id


MAX_TOTAL = 1_000_000
MAX_AMOUNT = 100_000
MAX_QUANTITY = 100
MIN_QUANTITY = 0

def validorder(order: Order):
    payments = Decimal('0.0')
    expenses = Decimal('0.0')

    for item in order.items:
        if item.type == 'payment':
            if -MAX_AMOUNT <= item.amount <= MAX_AMOUNT:
                payments += Decimal(str(item.amount))
        elif item.type == 'product':
            if type(item.quantity) is int and \
                    MIN_QUANTITY < item.quantity <= MAX_QUANTITY and \
                    MIN_QUANTITY < item.amount <= MAX_AMOUNT:
                expenses += Decimal(str(item.amount)) * item.quantity
        else:
            return "Invalid item type: %s" % item.type

    if abs(payments) > MAX_TOTAL or expenses > MAX_TOTAL:
        return "Total amount payable for an order exceeded"

    if payments != expenses:
        return "Order ID: %s - Payment imbalance: $%0.2f" % (order.id, payments - expenses)
    else:
        return "Order ID: %s - Full payment received!" % order.id

File: Season-2/Level-4/code.js
// Welcome to Secure Code Game Season-2/Level-4!

// Follow the instructions below to get started:

// 1. test.js is passing but the code here is vulnerable
// 2. Review the code. Can you spot the bugs(s)?
// 3. Fix the code.js but ensure that test.js passes
// 4. Run hack.js and if passing then CONGRATS!
// 5. If stuck then read the hint
// 6. Compare your solution with solution.js

const express = require("express");
const bodyParser = require("body-parser");
const libxmljs = require("libxmljs");
const multer = require("multer");
const path = require("path");
const fs = require("fs");
const { exec } = require("node:child_process");
const app = express();

app.use(bodyParser.json());
app.use(bodyParser.text({ type: "application/xml" }));

const storage = multer.memoryStorage();
const upload = multer({ storage });

app.post("/ufo/upload", upload.single("file"), (req, res) => {
  if (!req.file) {
    return res.status(400).send("No file uploaded.");
  }

  console.log("Received uploaded file:", req.file.originalname);

  const uploadedFilePath = path.join(__dirname, req.file.originalname);
  fs.writeFileSync(uploadedFilePath, req.file.buffer);

  res.status(200).send("File uploaded successfully.");
});

app.post("/ufo", (req, res) => {
  const contentType = req.headers["content-type"];

  if (contentType === "application/json") {
    console.log("Received JSON data:", req.body);
    res.status(200).json({ ufo: "Received JSON data from an unknown planet." });
  } else if (contentType === "application/xml") {
    try {
      // Fix 1, 2, 3
      // まずnonetオプションをtrueにすることで、
      // 解析時にドキュメントタイプ定義(<!DOCTYPE...>)を無効にする
      // そうすると、カスタムエンティティも無効になる
      //
      // また、この際recoverもfalseに設定する
      // hackの方で求められているのは400のErrorなので、
      // ここでエラーが起きるようにしたい
      //
      // なおエラーの原理としては、XML本文で`&xxe`を指定しているのに対し、
      // nonet=trueのおかげでカスタムエンティティ`xxe`がなく解析エラーが起きるため
      //
      // Fix 4
      // replaceEntitiesもfalseにする必要があった
      // nonet=trueでもreplaceEntitiesにより、
      // XXEのカスタムエンティティの参照先ファイルの中身を読み込んで取り替えてしまう
      // 今回のケースでは、読み込んだ内容がshellコマンドで、
      // 後のコマンド実行機能で任意のコマンド結果を漏洩する形となった
      const xmlDoc = libxmljs.parseXml(req.body, {
        // replaceEntities: true,
        // recover: true,
        // nonet: false,
        replaceEntities: false,
        recover: false,
        nonet: true,
      });

      console.log("Received XML data from XMLon:", xmlDoc.toString());

      const extractedContent = [];

      xmlDoc
        .root()
        .childNodes()
        .forEach((node) => {
          if (node.type() === "element") {
            extractedContent.push(node.text());
          }
        });

      // Secret feature to allow an "admin" to execute commands
      if (
        xmlDoc.toString().includes('SYSTEM "') &&
        xmlDoc.toString().includes(".admin")
      ) {
        extractedContent.forEach((command) => {
          exec(command, (err, output) => {
            if (err) {
              console.error("could not execute command: ", err);
              return;
            }
            console.log("Output: \n", output);
            res.status(200).set("Content-Type", "text/plain").send(output);
          });
        });
      } else {
        res
          .status(200)
          .set("Content-Type", "text/plain")
          .send(extractedContent.join(" "));
      }
    } catch (error) {
      console.error("XML parsing or validation error:", error.message);
      res.status(400).send("Invalid XML: " + error.message);
    }
  } else {
    res.status(405).send("Unsupported content type");
  }
});

const PORT = process.env.PORT || 3000;
const server = app.listen(PORT, () => {
  console.log(`Server is running on port ${PORT}`);
});

module.exports = server;
